### ISSUE TITLE
Files disappearing from search results

### ISSUE THREAD

**Issue by caseychoiniere** [Duke-Translational-Bioinformatics/duke-data-service#1073]
**Dev**

Files that were showing up in `POST/search/folders_files` search results suddenly stop showing up even though they haven't been deleted and the query is the same. For example, I created some files to test with search and they were showing up in results and then all of the sudden, 5 minutes later they disappeared from results. At first I thought it was an issue with my query but it's not.

Here's an example file id of an affected file `5ab0b5ae-4342-496f-8128-795505b2ebb6`

**Comment by dmlond** [comment]
Can you provide some more details of the exact actions that you took in this file's life?
I am having trouble reproducing the issue. I can state that for some reason 2 of the 5 data_files in the project containing the example file that you gave us (including the example) either
- never indexed in elasticsearch in the first place
- had their elasticsearch indices updated incorrectly
- were physically deleted from elasticsearch.

we do not have any code callbacks that delete, so I suspect either the first or second option. I would need some more information to try to reproduce.

**Comment by dmlond** [comment]
also, please attach the query you are using :)

**Comment by caseychoiniere** [comment]
1.) Created files. 
2.) Searched for files and files were discoverable.
3.) Searched for files and suddenly they were not discoverable.
4.) Came back to this today and without changing the query at all the file `5ab0b5ae-4342-496f-8128-795505b2ebb6` is discoverable again. I even checked my VCS to confirm that no changes were made to the query. It is possible that I supplied an incorrect file id here. It seems that all of the files that exist in project `484c88e0-fe2b-4e9b-901f-55b822713418` do not exist in Elasticsearch. One of these files (id `bfa6aa1a-5e80-4c32-a982-a9bd6d1df2ef`) has the same name as file `5ab0b5ae-4342-496f-8128-795505b2ebb6` which exists in a different project. 

5.) Files `0892a63d-ebf1-4e9c-85aa-a98e83544970`, `1bf57196-7a48-429a-ad54-37c0685d5654`, `0892a63d-ebf1-4e9c-85aa-a98e83544970` and `bfa6aa1a-5e80-4c32-a982-a9bd6d1df2ef` are not discoverable.

So it seems the issue here may be that files were not indexed in the first place. 

Query was as follows (query string === exact file name with no post filters applied) 

```javascript
{
            "query_string": {
                "query": query
            },
            "filters": [
                {"kind": kind}
            ],
            "aggs": [
                {
                    "field": "project.name",
                    "name": "project_names",
                    "size": 20
                },
                {
                    "field": "tags.label",
                    "name": "tags",
                    "size": 20
                },
            ],
            "post_filters": postFilters
        };
```

**Comment by dmlond** [comment]
no, its a certainty that you were not seeing that file in the response at some point. Sorry, I wasnt clear. I just need to see if I can diagnose why that and another file were not indexed in elasticsearch during a period that seems to have spanned 5 minutes after you created them to today.

**Comment by dmlond** [comment]
I am going to provisionally say that this may have been due to a transient glitch associated with the software as it existed from August 1 - September 5. I was able to diagnose an elasticsearch query (see below) to find about 496 records, all created between these dates, that had very incomplete elasticsearch index documents. They basically did not contain the kind or name fields that we allow to be queried with the mapping. This prevented them from being returned in any elasticsearch queries (even a match_all!!).  I was able to use the Ruby elasticsearch gem bindings to find the actual DB model objects affected, and index them properly. The above (and others) should now be returned in search.

Here is the elasticsearch query to use to find objects with incomplete documents.
```
GET /data_files/_search
{
    "query": {
      "bool": {
         "must_not": {
            "exists": {
                "field": "name"
            }
        }
      }
    }
}
```

Here is some Ruby code that I can run in the heroku rails console
```
query = {:query=>{:bool=>{:must_not=>{:exists=>{:field=>"kind"}}}}}
er = Elasticsearch::Model.search(query, [DataFile,Folder])
er.per(500)
#find minimum and maximum creation_date of the bad records
er.page(1).records.collect {|c| c.created_at }.minmax
er.page(1).records.each do |r|
  r.__elasticsearch__.index_document
end
```

**Comment by dmlond** [comment]
didnt meant to close and comment, just comment :)
